parlai/core/torch_classifier_agent.py

```diff
@@ -0,0 +1,285 @@
+        return shared
+
+    def _get_labels(self, batch):
+        """Obtain the correct labels. Raise an exception if one of the labels
+        is not in the class list.
+        """
+        try:
+            labels_indices_list = [self.class_dict[label] for label in
+                                   batch.labels]
+        except KeyError as e:
+            print('One of your labels is not in the class list.')
+            raise e
+        labels_tensor = torch.LongTensor(labels_indices_list)
+        if self.use_cuda:
+            labels_tensor = labels_tensor.cuda()
+        return labels_tensor
+
+    def _update_confusion_matrix(self, batch, predictions):
+        """Update the confusion matrix given the batch and predictions.
+
+        :param batch:         a Batch object (defined in torch_agent.py)
+        :param predictions:   (list of string of length batchsize) label
+                              predicted by the classifier
+        """
+        for i, pred in enumerate(predictions):
+            label = batch.labels[i]
+            self.metrics['confusion_matrix'][(label, pred)] += 1
+
+    def train_step(self, batch):
+        """Train on a single batch of examples."""
+        if batch.text_vec is None:
+            return Output()
+        self.model.train()
+        self.optimizer.zero_grad()
+
+        # calculate loss
+        labels = self._get_labels(batch)
+        scores = self.score(batch)
+        loss = self.classifier_loss(scores, labels)
+        loss.backward()
+        self.update_params()
+
+        # update metrics
+        self.metrics['loss'] += loss.item()
+        self.metrics['examples'] += len(batch.text_vec)
+
+        # get predictions
+        _, prediction_id = torch.max(scores.cpu(), 1)
+        preds = [self.class_list[idx] for idx in prediction_id]
+        self._update_confusion_matrix(batch, preds)
+
+        return Output(preds)
+
+    def eval_step(self, batch):
+        """Train on a single batch of examples."""
+        if batch.text_vec is None:
+            return
+
+        self.model.eval()
+        scores = self.score(batch)
+        probs = F.softmax(scores, dim=1)
+        if self.threshold is None:
+            _, prediction_id = torch.max(probs.cpu(), 1)
+        else:
+            ref_prob = probs.cpu()[:, 0]
+            # choose ref class if Prob(ref class) > threshold
+            prediction_id = ref_prob <= self.threshold
+        preds = [self.class_list[idx] for idx in prediction_id]
+
+        if self.opt.get('interactive_mode', False):
+            if self.opt.get('print_scores', False):
+                probabilities = []
+                for score in scores:
+                    probabilities.append(probs[0])
+                preds = []
+                for i, pred_id in enumerate(prediction_id):
+                    prob = round_sigfigs(probabilities[i][pred_id], 4)
+                    preds.append('{}: {}'.format(self.class_list[pred_id],
```

Comment:
I think this is strange behavior. Do you not want to give it back in a structured form?

Reply:
i'll clean this up